tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,290 @@
+        if atomic_number == ion_number:
+            rate_matrix_block[ion_number, :] = 1.0
+        else:
+            ion_rate_matrix = NLTERateEquationSolver.ion_matrix(
+                ion_rates, atomic_number
+            )
+            recomb_rate_matrix = NLTERateEquationSolver.recomb_matrix(
+                recomb_rate, atomic_number
+            )
+            rate_matrix_block[ion_number, :] = (
+                ion_rate_matrix + recomb_rate_matrix
+            )[ion_number, :]
+        return rate_matrix_block
+
+    @staticmethod
+    def lte_rate_matrix_block(phi_block, electron_density):
+        """Creates the generic LTE block for rate matrix.
+
+        Parameters
+        ----------
+        phi_block : DataFrame
+            Saha Factors for current atomic number
+        electron_density : float
+            Current guess for electron density
+
+        Returns
+        -------
+        numpy.array
+            LTE block for rate matrix
+        """
+        lte_rate_vector_block = -1.0 * np.hstack([*phi_block.values, -1.0])
+        lte_rate_matrix_block = np.diag(lte_rate_vector_block)
+        n_e_initial = np.ones(len(phi_block)) * electron_density
+        n_e_matrix = np.diag(n_e_initial, 1)
+        lte_rate_matrix_block += n_e_matrix
+        lte_rate_matrix_block[-1, :] = 1.0
+        return lte_rate_matrix_block
+
+    @staticmethod
+    def prepare_phi(phi):
+        phi[phi == 0.0] = 1.0e-10 * phi[phi > 0.0].min().min()
+        return phi
+
+    @staticmethod
+    def recomb_matrix(recomb_rate, atomic_number):
+        offdiag = np.zeros(atomic_number)
+        index = recomb_rate.index
+        for i in index:
+            offdiag[i] = recomb_rate.loc[i]
+        diag = np.hstack([np.zeros(1), -offdiag])
+        return np.diag(diag) + np.diag(offdiag, k=1)
+
+    @staticmethod
+    def ion_matrix(ion_rate, atomic_number):
+        offdiag = np.zeros(atomic_number)
+        index = ion_rate.index
+        for i in index:
+            offdiag[i] = ion_rate.loc[i]
+        diag = np.hstack([-offdiag, np.zeros(1)])
+        return np.diag(diag) + np.diag(offdiag, k=-1)
+
+    @staticmethod
+    def prepare_last_row(atomic_numbers):
+        last_row = []
+        for atomic_number in atomic_numbers:
+            last_row.append(np.arange(0.0, atomic_number + 1))
+        last_row = np.hstack([*last_row, -1])
+        # TODO needs to be modified for use in nlte_excitation
+        return last_row
+
+    @staticmethod
+    def prepare_ion_recomb_rates_nlte_ion(
+        gamma,
+        alpha_sp,
+        alpha_stim,
+        coll_ion_coeff,
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+    ):
+        indexer = pd.Series(
+            np.arange(partition_function.shape[0]),
+            index=partition_function.index,
+        )
+        _ion2level_idx = indexer.loc[levels.droplevel(2)].values
```

Comment:
Personally, I would prefer to use `level_number` instead of `2` in `droplevel` to be more explicit.

Reply:
Done